frame/contracts/src/tests.rs

```diff
@@ -1701,13 +1741,46 @@ fn chain_extension_works() {
 				0,
 				GAS_LIMIT,
 				None,
-				Input { extension_id: 2, func_id: 0, extra: &[] }.into(),
+				ExtensionInput { extension_id: 2, func_id: 0, extra: &[] }.into(),
 			),
 			Error::<Test>::NoChainExtension,
 		);
 	});
 }
 
+#[test]
+fn chain_extension_temp_storage_works() {
+	let (code, hash) = compile_module::<Test>("chain_extension_temp_storage").unwrap();
+	ExtBuilder::default().existential_deposit(50).build().execute_with(|| {
+		let min_balance = <Test as Config>::Currency::minimum_balance();
+		let _ = Balances::deposit_creating(&ALICE, 1000 * min_balance);
+		assert_ok!(Contracts::instantiate_with_code(
+			Origin::signed(ALICE),
+			min_balance * 100,
+			GAS_LIMIT,
+			None,
+			code,
+			vec![],
+			vec![],
+		),);
+		let addr = Contracts::contract_address(&ALICE, &hash, &[]);
+
+		// Call func 0 and func 1 back to back.
```

Comment:
How come you only test func_ids 2 and 3 with the handrolled Wasm fine instead of Rust unit tests?

Reply:
Because they need to be called in the same contract execution to test what they should be testing: The temporary buffer is preserved within the same call. You can't do that other than calling it from the same contract.